lib/elixir/lib/task/supervised.ex

```diff
@@ -377,20 +399,35 @@ defmodule Task.Supervised do
       # spawned processes (that are also linked) with the same reason, and then
       # exit ourself with the same reason.
       {:DOWN, ^parent_ref, _, _, reason} ->
-        for {_ref, {_counter, :link, pid}} <- counters do
+        for {_ref, {_position, :link, pid, _timer_ref}} <- running_tasks do
           Process.exit(pid, reason)
         end
         exit(reason)
 
       # One of the spawned processes went down. We inform the parent process of
       # this and keep going.
       {:DOWN, ref, _, _, reason} ->
-        {{counter, _, _}, counters} = Map.pop(counters, ref)
-        send(parent_pid, {:down, {monitor_ref, counter}, reason})
-        stream_monitor_loop(parent_pid, parent_ref, mfa, spawn, monitor_ref, counters)
+        {{position, _type, _pid, timer_ref}, running_tasks} = Map.pop(running_tasks, ref)
+        :ok = Process.cancel_timer(timer_ref, async: true, info: false)
+        send(parent_pid, {:down, {monitor_ref, position}, reason})
+        stream_monitor_loop(parent_pid, parent_ref, mfa, spawn, monitor_ref, running_tasks, timeout)
+
+      # One of the spawned processes timed out. We kill that process here
+      # regardless of the value of :on_timeout. We then send a message to the
+      # parent process informing it that a task timed out, and the parent
+      # process decides what to do.
+      {:timed_out, {^monitor_ref, ref}} ->
+        case running_tasks do
+          %{^ref => {position, _type, pid, _timer_ref}} ->
+            send(parent_pid, {:killed_for_timeout, {monitor_ref, position}})
+            Process.exit(pid, :kill)
```

Comment:
If we aren't trapping exits and `on_timeout: :exit` this creates 2 possibilities.

Either the task is killed and the exit signals propagates causing Caller to get `:killed` exit signal. Or the Caller handles the `:killed_for_timeout` message and tells the Monitor to stop all tasks. The Monitor receives the stop and then traps exits before task exits. Then the Caller will call `exit({:timeout, ..})`. This differing semantics is complex. I think we would need to `unlink` from the Monitor, perhaps similar to `Task.shutdown`.